src/coreComponents/meshUtilities/CPMesh/CPMeshParser.cpp

```diff
@@ -0,0 +1,372 @@
+  bool foundZCORN = false;
+  bool foundACTNUM = false;
+  bool foundPERMX = false;
+  bool foundPERMY = false;
+  bool foundPERMZ = false;
+
+  std::string line;
+  while( getline( meshFile, line ) )
+  {
+    CPMeshStringUtilities::removeStringAndFollowingContentFromLine( "--", line );
+    CPMeshStringUtilities::removeExtraSpaces( line );
+    CPMeshStringUtilities::removeEndOfLine( line );
+    CPMeshStringUtilities::removeTab( line );
+    CPMeshStringUtilities::trim( line );
+
+    // TODO: shorten, there must be a way to factorize all these functions
+    //       at least, we shoud be able to merge readLocalACTNUM and readLocalPROP
+
+    // at this point, SPECGRID has already been read, we skip it
+    if( line == "COORD" )
+    {
+      foundCOORD = true;
+      readLocalCOORD( meshFile, cPMeshDims );
+    }
+    else if( line == "ZCORN" )
+    {
+      foundZCORN = true;
+      readLocalZCORN( meshFile, cPMeshDims );
+    }
+    else if( line == "ACTNUM" )
+    {
+      foundACTNUM = true;
+      readLocalACTNUM( meshFile, cPMeshDims );
+    }
+    else if( line == "PERMX" )
+    {
+      foundPERMX = true;
+      readLocalPROP( meshFile, cPMeshDims, m_permx );
+    }
+    else if( line == "PERMY" )
+    {
+      foundPERMY = true;
+      readLocalPROP( meshFile, cPMeshDims, m_permy );
+    }
+    else if( line == "PERMZ" )
+    {
+      foundPERMZ = true;
+      readLocalPROP( meshFile, cPMeshDims, m_permz );
+    }
+    else if( line == "PORO" )
+    {
+      readLocalPROP( meshFile, cPMeshDims, m_poro );
+    }
+  }
+
+  if( !foundACTNUM )
+  {
+    fillLocalACTNUM( cPMeshDims );
+  }
+
+  // TODO: error out if one of them is missing
+  if( foundPERMX && foundPERMY && foundPERMZ )
+  {
+    m_perm.resizeDimension< 0, 1 >( m_permx.size(), 3 );
+    for( localIndex i = 0; i < m_permx.size(); ++i )
+    {
+      m_perm( i, 0 ) = m_permx( i );
+      m_perm( i, 1 ) = m_permy( i );
+      m_perm( i, 2 ) = m_permz( i );
+    }
+  }
+
+  GEOSX_THROW_IF( !foundCOORD || !foundZCORN,
+                  "At least one of the following keywords was not found: COORD, ZCORN",
+                  InputError );
+
+
+}
+
+void CPMeshParser::readLocalCOORD( std::istringstream & meshFile,
+                                   CPMeshDimensions const & cPMeshDims )
+{
+  localIndex const nX = cPMeshDims.nX();
+  localIndex const nY = cPMeshDims.nY();
+
+  localIndex const maxSize = 6*(nX+1)*(nY+1); // TODO: change
+  std::vector< real64 > work;
+  work.reserve( maxSize );
+
+  string buffer = extractDataBelowKeyword( meshFile ); // TODO: change to save only what is necessary
+  CPMeshStringUtilities::eclipseDataBufferToVector( buffer, work ); // TODO: change to save only what is necessary
+
+  localIndex const nXLocal = cPMeshDims.nXLocal();
+  localIndex const nYLocal = cPMeshDims.nYLocal();
+
+  localIndex const iMinLocal = cPMeshDims.iMinLocal();
+  localIndex const jMinLocal = cPMeshDims.jMinLocal();
+
+  m_coord.resize( 6*(nXLocal+1)*(nYLocal+1) );
```

Comment:
why 6? `m_coord` contains only x and y coordinates right?

Reply:
Here is an explanation from the [OPM documentation](https://opm-project.org/wp-content/uploads/2019/01/OPM-Flow-Documentation-2018-10-Rev-2.pdf): 
```
COORD defines a set of coordinate lines or pillars for a reservoir grid via an array. 
A total of (NX+1) x (NY+1) lines must be specified for each coordinate data set. 
For Cartesian geometry, each line is defined by the (x, y, z) coordinates of two distinct points on the line. 
```
For each of two points, you need the (x,y,z) coordinates, hence 6 entries per pillar.